### ISSUE TITLE
Applying custom functions on the data before parsing

### ISSUE THREAD

**Issue by Sree2015** [databricks/spark-csv#399]
I want to apply custom functions to clean the data before spark-csv starts to parse the data. Is there a way to do that with current versions?

**Comment by HyukjinKwon** [comment]
@Sree2015 Maybe, you could use [CsvParser.csvRdd(...)](https://github.com/databricks/spark-csv/blob/1ae649285462df1af1411593e2abe589de2d704c/src/main/scala/com/databricks/spark/csv/CsvParser.scala#L157) API. After loading it into `RDD[String]` and cleaning up the data, then you could use this API to load it into `DataFrame`.

**Comment by Sree2015** [comment]
CsvParser.csvRdd calls CsvRelation which parses the file. I want to do the cleaning before parsing.

**Comment by HyukjinKwon** [comment]
Could I please ask what you mean by "custom functions to clean the data"? I thought you meant you want some transformation before converting it into `DataFrame`.

**Comment by Sree2015** [comment]
For example for decimals we get signs at the end as they are treated as strings. I will have to write a custom function to move the sign to the beginning such that it can be parsed while being loaded into data frame with custom schema.

**Comment by falaki** [comment]
I suggest you load that field as string, and then apply a simple transformation (or UDF) to get a new DataFrame (or table) with desired column.

**Comment by Sree2015** [comment]
In this case i will have to use inferSchema option and then apply UDFs for specific columns and then load into another dataframe with custom schema. Is this right?

**Comment by falaki** [comment]
Yes. Or you can do everything with DataFrames to avoid the last step.

**Comment by Sree2015** [comment]
I would want to do everything within the same data frame, however the issue is if the data is not transformed before applying the custom schema, it fails parsing. Is there a way to overcome this and still do in single dataframe?
